[nibeheatpump] #5012, #5604 Channel update logic, to and from modbus data conversion fixes

yozik04 opened the pull request:
Update channels even if value did not change.

This lets us to use expire binding to null values if there is a problem with modbus link to the heatpump or configuration.

Current version updates only changed values. Expire would remove values from some variables that do not change. Which is bad.

Fixes #5012
Fixes #5604

wborn commented:
The binding should indeed still continue to push same state updates unless it results in performance issues. 

If it didn't do this because of performance issues, this wil most likely have been addressed by the PRs for https://github.com/eclipse/smarthome/issues/6416.

paulianttila commented:
This feature has been implemented by purpose. I propose to add configuration variable to thing configuration, which define the behaviour.

yozik04 commented:
What was the purpose?

wborn commented:
What is exactly the purpose @paulianttilla?

yozik04 commented:
Another option would be to use same throthling that you implemented. But binding should update channels periodically regardless of if value were changed or not.

paulianttila commented:
I have implemented several bindings which update channels/items rather often. If all updates are passed, in my environment i get over 30 item updates in avg per second. This cause many problems, e.g habpanel is useless. This is the reason why I implemented both throttle and update filtering features.

What is the use case for expire binding use in heat pump device?

wborn commented:
> This cause many problems, e.g habpanel is useless.

If that's the case it should be addressed in the framework or HABPanel instead of each binding. Most other bindings do send same state updates or throttle these to max 1 per second.

Is this a known HABPanel issue @ghys?

ghys commented:
If you get of lot of traffic on the SSE endpoint (30 events/s seems a lot to me) then I've heard you get performance problems especially on older hardware or cheap tablets, but it isn't really surprising and should affect all UIs which subscribe to `/rest/events` like Paper UI.

paulianttila commented:
> Most other bindings do send same state updates or throttle these to max 1 per second.

I don't understand why state should be updated if state has not been changed? For me this is useless update. I have over 1000 items on my system, so why system should be stressed about useless updates?

yozik04 commented:
> What is the use case for expire binding use in heat pump device?

I will explain what problem I am trying to solve. Couple days ago I had outage of orangepi that runs nibegw. iOS app showed everything is fine so I was not aware of it. So you look in the app and see that current boiler temperature is good but when you go to shower water ends. You look in the app and see that water in the boiler is still the same. Then you lunch you computer and start digging...

All my other bindings use `expire = "10m"`. So if there were not updates to the item for 10 minutes consider it's data outdated. Then I see in the iOS app that I have no value for an item, probably due to misconfiguration or an outage. I am not fooled with wrong value if something is wrong.

@paulianttila BTW, what RS485 module do you use? The one that worked for me for a month, now unable to read data from the heatpump. Checksum fails for long messages.

yozik04 commented:
@paulianttila Maybe you are ok if I do this way?
Using the same configuration.throttleTime ?
```
if (oldValue != null && val == oldValue.value && (oldValue.lastUpdateTime + configuration.throttleTime) < System.currentTimeMillis()) {
                logger.trace("Value did not change, ignoring update");
} else {
```

wborn commented:
> I don't understand why state should be updated if state has not been changed?

It's specifically mentioned in the [ESH docs](https://www.eclipse.org/smarthome/documentation/development/bindings/thing-handler.html#polling-for-a-state):

> Even if the state has not changed since the last update, the binding should inform the framework, because it indicates that the value is still present.

If you'd update an item using rules or a UI (similar to the expire binding) it will never get the current state again unless the state changes. Are there more reasons @kaikreuzer?

paulianttila commented:
>I will explain what problem I am trying to solve. Couple days ago I had outage of orangepi that runs nibegw. iOS app showed everything is fine so I was not aware of it. So you look in the app and see that current boiler temperature is good but when you go to shower water ends. You look in the app and see that water in the boiler is still the same. Then you lunch you computer and start digging...

>All my other bindings use expire = "10m". So if there were not updates to the item for 10 minutes consider it's data outdated. Then I see in the iOS app that I have no value for an item, probably due to misconfiguration or an outage. I am not fooled with wrong value if something is wrong.

I have defined one DateTime itemper critical devices (one for Nibe pump), which I update by rules. For Nibe devices I monitor degree minutes. If degree minutes have not updated in one hour, I send an alarm notification (email + push notification). 

Nibe binding could be improved to monitor communication, if any message haven't received in certain time, binding should but thing to offline state. Thing state changes can be monitored by rules as well. But if binding have jammed, thing state might not be updated either, so that's why I have done last update monitoring by DateTime item.

>what RS485 module do you use? 

I used cheap Arduino board solution several years with OH1 binding without problems. Nowadays I use ProDino board (arduino based, which have RS-485 on board), which I have mentation also in binding page. I have run ProDino few years now, without single problem.

paulianttila commented:
>If you'd update an item using rules or a UI (similar to the expire binding) it will never get the current state again unless the state changes.

By UI, you normally send commands not updates. By rules, item could be updated, but not sure why that should be done. Anyhow, item/channel update can be forced by sending REFRESH command.

davidgraeff commented:
`/rest/events` is actually the problem. This PR adapts the binding to behave as it should according to the coding guidelines. It doesn't matter if the framework does the `if changed` or the binding is doing it btw.

The framework unfortunately pushes everything to the mentioned http endpoint, which it shouldn't do without being asked to do so.

wborn commented:
> I proposed that this could be configurable on the binding 

If we make this configurable for this binding it sets a precedent for adding similar configurations to other bindings. So if we can prevent this by addressing the root cause we reduce the overall complexity.

J-N-K commented:
> If you'd update an item using rules or a UI (similar to the expire binding) it will never get the current state again unless the state changes. Are there more reasons @kaikreuzer?

For persistence it may be necessary to know the age of a value. Something like `expire` will fail if no updates are send.

kaikreuzer commented:
> Are there more reasons @kaikreuzer?

As the docs say: Whenever we have knowledge about the "real" (external) state, we should sync this information to the internal state. And yes, the event happens at a certain time, so the important information is the moment of the event - we know that this state is pretty recent and not 2 days old.

We specifically introduced the separation between StateUpdated and StateChanged events. Afaik, the REST endpoint should actually swallow the updated events and only send the changed events out to subscribers. So in theory, the only bottleneck should be the internal event bus and if some bindings polls data with high frequency, some throttling might make sense (as for HA use cases, it usually does not matter whether a state is 1 second old or only 300ms).

yozik04 commented:
So what we do with that?
I see that file location has changed. If I will move my changes to a new location, what will happen next. Do I need to try to implement some caching mechanism?

The heatpump publishes readouts of all configured channels every 2-3 seconds.

Solutions:
- As in PR. All updates from the bus trigger updateState regardless value changed or not. 
- Implement a throttling mechanism. If a channel value was not changed for a minute(configurable) then trigger updateState?

davidgraeff commented:
Usually you cache values in a binding to make them available to the framework asap when it requests them. When you have new values for a channel, no matter if changed or not, just pass them to the framework. Throttling is task of the core.

yozik04 commented:
The Core logs every time... In last commit I've used same cache as was already implemented in the binding. Should be a better solution that makes everybody happy.

Also reviewed Arduino NibeGW code, it had some bugs:
- It stoped responding after first write/read command.
- Added packet flushing as Arduino library recommends.
- Added support of ENC82j60. With fully enabled debug it does not fit into the memory. PROGMEM???

RPi NibeGW version does not work for me. Produces a lot of checksum failures and heatpump goes to safe mode. But it is problem of RPi UART handling in the kernel. Tried 3x RS485 modules, no luck.

With new structure I am unable to build package locally anymore...

I do not develop in Java every day. I have no idea how to fix it. All this local build setup is too time consuming...

```
[INFO] Scanning for projects...
[INFO]
[INFO] ----< org.openhab.addons.bundles:org.openhab.binding.nibeheatpump >-----
[INFO] Building openHAB Add-ons :: Bundles :: Nibe Heatpump Binding 2.5.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-enforcer-plugin:3.0.0-M2:enforce (enforce-java) @ org.openhab.binding.nibeheatpump ---
[WARNING] Rule 0: org.apache.maven.plugins.enforcer.RequireJavaVersion failed with message:
Detected JDK Version: 1.8.0-11 is not in the allowed range [1.8.0-40,1.9),[9.0,12.0).
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.013 s
[INFO] Finished at: 2019-05-14T21:57:03+03:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M2:enforce (enforce-java) on project org.openhab.binding.nibeheatpump: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed. -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

davidgraeff commented:
> Detected JDK Version: 1.8.0-11 is not in the allowed range [1.8.0-40,1.9),[9.0,12.0).

But that's obvious. Your java installation is too old.

> The Core logs every time

I know. And that's the most ~~stupid~~ not-so-well-engineered part that was ever committed to openHAB IMO. It destroys sd-cards and is unnecessary:
* If someone wants a historical event recording, he'd use the persistence API and not the log file.
* If someone wants to see if an item changed, he'd look at the item state itself.
* If he wants to know if items are updated, he'd look at the event-bus for example via the karaf console.

@kaikreuzer Can we please turn off event logging completely by removing the bundle that logs item events? (As far as I remember there is a dedicated bundle or at least package doing that)

yozik04 commented:
Ahh. It was written in the log.... 🤦‍♂ 
`[INFO] BUILD SUCCESS`

Will test it as soon as my heatpump will be maintained after ground pipe cut... 😞 

Logging to a file is also not a great idea. I'd use syslog. But that is another topic...

paulianttila commented on bundles/org.openhab.binding.nibeheatpump/contrib/NibeGW/Arduino/NibeGW/NibeGW.ino:
```diff
@@ -234,24 +245,37 @@ void nibeCallbackMsgReceived(const byte* const data, int len)
+      int packetSize = udp4readCmnds.parsePacket();
+      if (packetSize) {
+        len = udp4readCmnds.read(data, packetSize);
+        udp4readCmnds.flush();
```
Flush applies only for output, right? So why it's needed here?

yozik04 replied:
You are right. It is not required for Ethernet library and only required for UIPEthernet (ENC28J60).

yozik04 replied:
Fixed it.

paulianttila commented on bundles/org.openhab.binding.nibeheatpump/src/main/java/org/openhab/binding/nibeheatpump/internal/handler/NibeHeatPumpHandler.java:
```diff
@@ -411,7 +411,7 @@ private void parseWriteEnabledItems() throws IllegalArgumentException {
     }
 
-    private State convertNibeValueToState(NibeDataType dataType, double value, String acceptedItemType) {
+    private State convertNibeValueToState(VariableInformation variableInfo, int value, String acceptedItemType) {
```
Maybe we should add junit tests for this. WDYT?

yozik04 replied:
I actually did a test to verify this code, but method is private and have no idea how to test it without making it at least protected.

yozik04 replied:
Reflection helped =) Learning every day.

yozik04 commented:
It is so damn hard to convert bytes to proper types in Java... It is so simple in python...

In java if value was casted to (short) before assignment to an int variable it will produce totally different result...

yozik04 commented:
Now it just can't be wrong. Fully rebuilding values from byte array.

wborn commented on bundles/org.openhab.binding.nibeheatpump/src/test/java/org/openhab/binding/nibeheatpump/internal/handler/NibeHeatPumpHandlerCommand2NibeTest.java:
```diff
@@ -0,0 +1,95 @@
+/**
+ * Tests cases for {@link NibeHeatPumpHandler}.
+ *
+ * @author Jevgeni Kiski
```
```suggestion
 * @author Jevgeni Kiski - Initial contribution
```

wborn commented on bundles/org.openhab.binding.nibeheatpump/src/test/java/org/openhab/binding/nibeheatpump/internal/handler/NibeHeatPumpHandlerNibe2StateTest.java:
```diff
@@ -0,0 +1,97 @@
+/**
+ * Tests cases for {@link NibeHeatPumpHandler}.
+ *
+ * @author Jevgeni Kiski
```
```suggestion
 * @author Jevgeni Kiski - Initial contribution
```

yozik04 replied:
Will add in next commits. Thank you.

yozik04 commented:
Got my heatpump pipes repaired and want to finish this PR.
Finally I like how the binding works now. Requested reads are now also properly parsed.
Heatpump sends some junk with payload if data is uInt_8. 2424832 in payload actually equals 0. Parsing fixed.

Should I squash all commits into one?

yozik04 commented:
Some more bugs to fix. Requested read still has some bugs 😞:
```
2019-05-17 23:55:53.134 [vent.ItemStateChangedEvent] - Nibe_1255_Total_Electrical_Addition_Operation_Time changed from -6.0 h to 19.6 h
2019-05-17 23:55:53.137 [vent.ItemStateChangedEvent] - Nibe_1255_Operation_Time_Compr changed from 9487 to 9487.0 h
2019-05-17 23:55:53.137 [vent.ItemStateChangedEvent] - Nibe_1255_Operation_Time_Compr_Hot_Water changed from -107 to 1685.0 h
2019-05-17 23:55:53.139 [vent.ItemStateChangedEvent] - Nibe_1255_Compressor_Starts changed from -7.0 to 4857.0
2019-05-17 23:55:59.456 [vent.ItemStateChangedEvent] - Nibe_1255_Total_Electrical_Addition_Operation_Time changed from 19.6 h to -6.0 h
2019-05-17 23:56:06.350 [vent.ItemStateChangedEvent] - Nibe_1255_Operation_Time_Compr_Hot_Water changed from 1685.0 h to -107
```

yozik04 commented:
Fixed as well.

paulianttila commented on bundles/org.openhab.binding.nibeheatpump/src/main/java/org/openhab/binding/nibeheatpump/internal/handler/NibeHeatPumpHandler.java:
```diff
@@ -411,29 +422,67 @@ private void parseWriteEnabledItems() throws IllegalArgumentException {
+        int decimals = (int)Math.log10(variableInfo.factor);
+        byte[] bytes;
+        int x;
+        switch (dataType) {
```
Your conversions seems rather complex to me, especially those signed values :)

Should following do the same (I didn't do any tests):

long v;

switch (dataType) {
	case U8:
		v = (byte) value & 0xFF;
		break;
	case U16:
		v = (short) value & 0xFFFF;
		break;
	case U32:
		v = value & 0xFFFFFFFFL;
		break;
	case S8:
		v = (byte) value;
		break;
	case S16:
		v = (short) value;
		break;
	case S32:
		v = value;
		break;
}

BigDecimal converted = new BigDecimal(v).movePointLeft(decimals).setScale(decimals, RoundingMode.HALF_EVEN);

yozik04 replied:
That will be too simple that everybody would understand. 😄 

Found another way. Much simpler.

yozik04 commented on bundles/org.openhab.binding.nibeheatpump/contrib/NibeGW/Arduino/NibeGW/NibeGW.ino:
```diff
@@ -48,6 +48,9 @@
 #define ETH_INIT_DELAY          10
 
+// Enable if ENC28J60 LAN module is used
+//#define TRANSPORT_ETH_ENC28J60
```
done

yozik04 commented on bundles/org.openhab.binding.nibeheatpump/src/main/java/org/openhab/binding/nibeheatpump/internal/handler/NibeHeatPumpHandler.java:
```diff
@@ -421,19 +429,53 @@ private State convertNibeValueToState(NibeDataType dataType, double value, Strin
             switch (dataType) {
                 case U8:
+                    bytes = ByteBuffer.allocate(1).put((byte)value).array();
+                    value = bytes[0];
```
Nope. `value = (byte)(value & 0xFF)` will do.

yozik04 commented on bundles/org.openhab.binding.nibeheatpump/src/test/java/org/openhab/binding/nibeheatpump/internal/handler/NibeHeatPumpHandlerNibe2StateTest.java:
```diff
@@ -0,0 +1,110 @@
+                { 40940, (int)((short)0xFFFF), "%.1f", "Number", "-0.1"}, // these tests pass, but is it good?
+                { 40940, (short)0xFFFF, "%.1f", "Number", "-0.1"}, // these tests pass, but is it good?
+                { 40940, (byte)0xFFFF, "%.1f", "Number", "-0.1"}, // these tests pass, but is it good?
+                { 40940, (int)(0xFFFF), "%.1f", "Number", "6553.5"}, // these tests pass, but is it good?
```
With simple value convert approach I do not like this: Why produced result depends on what was initial cast?

yozik04 replied:
For S32:
This might be a solution:
`x = (int)((long)value << 1) >>> 1;`

Any ideas?

paulianttila commented on bundles/org.openhab.binding.nibeheatpump/contrib/NibeGW/Arduino/NibeGW/NibeGW.ino:
```diff
@@ -66,15 +68,21 @@
 #endif
 
+#ifndef TRANSPORT_ETH_ENC28J60
 #include <EthernetUdp.h>
```
Could this be moved to earlier TRANSPORT_ETH_ENC28J60

yozik04 replied:
Moved.

paulianttila commented on bundles/org.openhab.binding.nibeheatpump/src/main/java/org/openhab/binding/nibeheatpump/internal/handler/NibeHeatPumpHandler.java:
```diff
@@ -533,25 +514,49 @@ private void handleVariableUpdate(PumpModel pumpModel, ModbusValue value) {
-            stateMap.put(coilAddress, new CacheObject(System.currentTimeMillis(), val));
 
-            if (oldValue != null && val == oldValue.value) {
+            if (oldValue != null && val == oldValue.value && (oldValue.lastUpdateTime + refreshIntervalMillis() / 2) >= System.currentTimeMillis()) {
```
Why refresh interval is diveded by 2?

yozik04 replied:
If I set it to just `oldValue.lastUpdateTime + refreshIntervalMillis()` then binging requests to read variables one by one that are already included in the readout. We do not want that.

BTW. If we do `updateState `on every readout. Nothing is logged to the logs. Core just filters these events out. So @wborn is totally right.

wborn replied:
Thanks for also checking it! I also retested this last week and didn't see same state updates being sent to frontends using SSE.

Previously channel updates were expensive operations but this bottleneck was fixed with several PRs for https://github.com/eclipse/smarthome/issues/6416. It made channel updates for each binding a lot faster and also allowed for removing update optimizations for each binding. I removed mine from the Nest Binding in https://github.com/openhab/openhab2-addons/pull/4203. :-)

yozik04 replied:
@paulianttila So how you want it?

yozik04 replied:
@paulianttila Are you fine with current version?

paulianttila replied:
I just read this post on the [forum](https://community.openhab.org/t/your-setup-for-a-wall-mounted-tablet/22540/146) where this guy implemented a filter to rest interface and modified habpanel. So that put me in doubt if this problem is really solved.

yozik04 replied:
In that post he speaks about problem of initial load of ALL items to habpanel. Not about periodic updates.

What `refreshInterval` you personally use?
Why it is bad to send just **2** updates to unchanged channels (read from readout) during that period?

yozik04 replied:
I am running version of this pull request of the binding from end of May. I do not see any lags anywhere. Updates are minimal. events.log get updated only for changes. I control Point Offset with it, depending on the indoor climate, sun activity and weather forecast.
`expires` works perfectly. If I turn off the heat pump for ten minutes or more then all items get nullified... Perfect!

J-N-K commented:
@wborn, any objections on merging this?

yozik04 commented:
@paulianttila does not respond since Jul 10...

paulianttila commented:
Can be merged from my point of view. I try to make some test before 2.5.

wborn commented:
Thanks for your feedback @paulianttila!

yozik04 commented:
Thank you @paulianttila !
